File: NewsDaily/texttoaudiofile/src/main/java/com/andruid/magic/texttoaudiofile/api/TtsApi.java
package com.andruid.magic.texttoaudiofile.api;

import android.content.Context;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.widget.Toast;

import com.andruid.magic.texttoaudiofile.util.FileUtils;

import java.io.File;
import java.util.Locale;

import timber.log.Timber;

import static com.andruid.magic.texttoaudiofile.data.Constants.DIR_TTS;

public class TtsApi {
    private static TtsApi sInstance;
    private static final Object LOCK = new Object();
    private static TextToSpeech tts;
    private static File dir;
    private static boolean ttsInit = false;

    public static void init(Context context){
        dir = new File(context.getCacheDir(), DIR_TTS);
        initTTS(context);
    }

    public static TtsApi getInstance(){
        if(sInstance == null){
            synchronized (LOCK){
                Timber.d("ttsapi instance created");
                sInstance = new TtsApi();
            }
        }
        return sInstance;
    }

    private static void initTTS(final Context context) {
        tts = new TextToSpeech(context, new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {
                if(status == TextToSpeech.SUCCESS){
                    int result = tts.setLanguage(Locale.getDefault());
                    if(result == TextToSpeech.LANG_MISSING_DATA || result == TextToSpeech.LANG_NOT_SUPPORTED)
                        Toast.makeText(context, "Text to speech not available", Toast.LENGTH_SHORT).show();
                    else
                        ttsInit = true;
                }
            }
        });
    }

    public void convertToAudioFile(String text, String utteranceId,
                                   final AudioConversionListener mListener){
        if(!dir.exists()) {
            boolean res = dir.mkdir();
            Timber.d("dir created %s", res);
        }
        if(ttsInit) {
            tts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                @Override
                public void onStart(String utteranceId) {}

                @Override
                public void onDone(String utteranceId) {
                    File file = new File(dir, FileUtils.getFileName(utteranceId));
                    mListener.onAudioCreated(file);
                }

                @Override
                public void onError(String utteranceId) {
                    mListener.onFailure("Failed creating audio for utterance id "+utteranceId);
                }
            });
            File file = new File(dir, FileUtils.getFileName(utteranceId));
            tts.synthesizeToFile(text, null, file, utteranceId);
        }
        else
            mListener.onFailure("TTS init failed");
    }

    public boolean isReady() {
        return ttsInit;
    }

    public static void release(){
        tts.shutdown();
    }

    public interface AudioConversionListener {
        void onAudioCreated(File file);
        void onFailure(String msg);
    }
}

File: NewsDaily/texttoaudiofile/src/main/java/com/andruid/magic/texttoaudiofile/api/TtsApi.kt
package com.andruid.magic.texttoaudiofile.api

import android.app.Application
import android.content.Context
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import com.andruid.magic.texttoaudiofile.data.DIR_TTS
import com.andruid.magic.texttoaudiofile.data.model.TtsResult
import com.andruid.magic.texttoaudiofile.util.FileUtils.getFileName
import java.io.File
import java.util.*
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

object TtsApi {
    private val TAG = TtsApi::class.java.simpleName

    private lateinit var tts: TextToSpeech
    private lateinit var dir: File

    var isReady = false
        private set

    fun init(application: Application) {
        dir = File(application.cacheDir, DIR_TTS)
        initTTS(application.applicationContext)
    }

    fun release() {
        tts.shutdown()
    }

    private fun initTTS(context: Context) {
        tts = TextToSpeech(context, TextToSpeech.OnInitListener {
            val status = it
            if (status == TextToSpeech.SUCCESS) {
                val result = tts.setLanguage(Locale.getDefault())
                if (result != TextToSpeech.LANG_MISSING_DATA && result != TextToSpeech.LANG_NOT_SUPPORTED)
                    isReady = true
            }
        })
    }

    suspend fun convertToAudioFile(text: String, utteranceId: String): TtsResult {
        lateinit var result: Continuation<TtsResult>

        if (!dir.exists()) {
            val res = dir.mkdir()
            Log.d(TAG, "dir created $res")
        }

        if (isReady) {
            tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onDone(utteranceId: String) {
                    val file = File(dir, getFileName(utteranceId))
                    result.resume(TtsResult.Success(file))
                }

                override fun onError(utteranceId: String) {
                    result.resume(TtsResult.Error("Failed creating audio for utterance id $utteranceId"))
                }

                override fun onStart(utteranceId: String) {}
            })

            val file = File(dir, getFileName(utteranceId))
            tts.synthesizeToFile(text, null, file, utteranceId)
        } else
            result.resume(TtsResult.Error("TTS init failed"))

        return suspendCoroutine { continuation -> result = continuation }
    }
}